modules/apps/portal-search/portal-search/src/main/java/com/liferay/portal/search/internal/buffer/IndexerRequestBufferExecutorWatcher.java

```diff
@@ -64,44 +62,23 @@ public IndexerRequestBufferExecutor getIndexerRequestBufferExecutor() {
 
 	@Activate
 	@Modified
-	protected void activate(Map<String, Object> properties) {
+	protected void activate(
+		BundleContext bundleContext, Map<String, Object> properties) {
+
 		_indexerRegistryConfiguration = ConfigurableUtil.createConfigurable(
 			IndexerRegistryConfiguration.class, properties);
-	}
-
-	@Reference(
-		cardinality = ReferenceCardinality.MULTIPLE,
-		policy = ReferencePolicy.DYNAMIC,
-		policyOption = ReferencePolicyOption.GREEDY
-	)
-	protected void addIndexerRequestBufferExecutor(
-		IndexerRequestBufferExecutor indexerRequestBufferExecutor,
-		Map<String, Object> properties) {
-
-		_indexerRequestBufferExecutors.put(
-			_getBufferedExecutionMode(properties),
-			indexerRequestBufferExecutor);
-	}
 
-	protected void removeIndexerRequestBufferExecutor(
-		IndexerRequestBufferExecutor indexerRequestBufferExecutor,
-		Map<String, Object> properties) {
-
-		_indexerRequestBufferExecutors.remove(
-			_getBufferedExecutionMode(properties));
+		_serviceTrackerMap = ServiceTrackerMapFactory.openSingleValueMap(
+			bundleContext, IndexerRequestBufferExecutor.class, null,
+			(indexerRequestBufferExecutor, emitter) -> emitter.emit(
+				GetterUtil.getString(
+					indexerRequestBufferExecutor.getProperty(
+						"buffered.execution.mode"))));
 	}
 
-	private String _getBufferedExecutionMode(Map<String, Object> properties) {
-		String bufferedExecutionMode = GetterUtil.getString(
-			properties.get("buffered.execution.mode"));
-
-		if (Validator.isNull(bufferedExecutionMode)) {
```

Comment:
@tinatian We've seen several cases where in the reference methods that maps a property value to a service, the method logic will throw exception if the property is not available on the bound service. They choose to do this instead of just filtering out services using `target` in `@Reference`. I wonder if this is really necessary -- is this just for logging, or reminding developers to remember to add the property?

When we switch it to `ServiceTrackerMap` or even `ServiceTracker + Map`, do we need to keep it?